src/serializer/ResidualHeapGraphGenerator.js

```diff
@@ -0,0 +1,149 @@
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+/* @flow */
+
+import type { Logger } from "./logger.js";
+import type { Modules } from "./modules.js";
+import type { Realm } from "../realm.js";
+import type { ObjectRefCount, AdditionalFunctionEffects } from "./types.js";
+
+import invariant from "../invariant.js";
+import {
+  Value,
+  EmptyValue,
+  FunctionValue,
+  AbstractValue,
+  SymbolValue,
+  ProxyValue,
+  ObjectValue,
+} from "../values/index.js";
+import { ResidualHeapInspector } from "./ResidualHeapInspector.js";
+import { ResidualHeapVisitor } from "./ResidualHeapVisitor.js";
+
+type Edge = {
+  fromId: number,
+  toId: number,
+};
+
+/**
+ * Generate a visualizable objects graph for Prepack heap.
+ */
+export class ResidualHeapGraphGenerator extends ResidualHeapVisitor {
+  constructor(
+    realm: Realm,
+    logger: Logger,
+    modules: Modules,
+    additionalFunctionValuesAndEffects: Map<FunctionValue, AdditionalFunctionEffects>,
+    valueToEdgeRecord: Map<Value, ObjectRefCount>
+  ) {
+    super(realm, logger, modules, additionalFunctionValuesAndEffects);
+    this._valueToEdgeRecord = valueToEdgeRecord;
+    this._visitedValues = new Set();
+    this._valueIds = new Map();
+    this._idSeed = 0;
+    this._path = [];
+    this._edges = [];
+  }
+
+  _valueToEdgeRecord: Map<Value, ObjectRefCount>;
+  _valueIds: Map<Value, number>;
+  _idSeed: number;
+  _visitedValues: Set<Value>;
+  _path: Array<Value>; // Contains the path of nodes from root to current visiting node.
+  _edges: Array<Edge>;
+
+  // Override.
+  preProcessValue(val: Value): boolean {
+    if (this._shouldIgnore(val)) {
+      return true;
+    }
+    this._updateEdge(val);
+
+    if (this._visitedValues.has(val)) {
+      return false; // Already visited.
+    }
+    this._visitedValues.add(val);
+    return true;
+  }
+
+  // Override.
+  postProcessValue(val: Value): void {
+    if (this._shouldIgnore(val)) {
+      return;
+    }
+    invariant(this._path.length > 0);
+    this._path.pop();
+  }
+
+  _getValueId(val: Value): number {
+    let id = this._valueIds.get(val);
+    if (!id) {
+      this._valueIds.set(val, ++this._idSeed);
+      id = this._idSeed;
+    }
+    return id;
+  }
+
+  _shouldIgnore(val: Value): boolean {
+    return val instanceof EmptyValue || val.isIntrinsic() || ResidualHeapInspector.isLeaf(val);
+  }
+
+  _updateEdge(val: Value) {
+    if (this._path.length > 0) {
+      const parent = this._path[this._path.length - 1];
+      this._edges.push({ fromId: this._getValueId(parent), toId: this._getValueId(val) });
+    }
+    this._path.push(val);
+  }
+
+  _getValueLabel(val: Value): string {
+    // TODO: add object identifier in label.
+    // TODO: does not use ref count yet, figure out how to best visualize it later.
+    return val.__originalName || "";
+  }
+
+  _generateDotGraph(nodes: Set<Value>, edges: Array<Edge>): string {
+    let content = "digraph{\n";
+
```

Comment:
Can you add comment listing out all the information you would like to display on the UI for nodes and edges. I will look into designing the UI so they can be display when they are available here.